v2/internal/testcommon/vcr/v3/replay_roundtripper.go

```diff
@@ -0,0 +1,153 @@
+		puts:  make(map[string]*http.Response),
+		log:   log,
+	}
+}
+
+// RoundTrip implements http.RoundTripper.
+func (replayer *replayRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	if request.Method == http.MethodGet {
+		return replayer.roundTripGet(request)
+	}
+
+	if request.Method == http.MethodPut {
+		return replayer.roundTripPut(request)
+	}
+
+	// For other kinds of request, just pass through to the inner round tripper.
+	return replayer.inner.RoundTrip(request)
+}
+
+func (replayer *replayRoundTripper) roundTripGet(request *http.Request) (*http.Response, error) {
+	// First use our inner round tripper to get the response.
+	response, err := replayer.inner.RoundTrip(request)
+	if err != nil {
+		// We have an error - return it, unless it's from go-vcr
+		if !errors.Is(err, cassette.ErrInteractionNotFound) {
+			return response, err
+		}
+
+		replayer.padlock.Lock()
+		defer replayer.padlock.Unlock()
+
+		// We didn't find an interaction, see if we have a cached response to return
+		if cachedResponse, ok := replayer.gets[request.URL.String()]; ok {
+			replayer.log.Info("Replaying GET request", "url", request.URL.String())
+			return cachedResponse, nil
+		}
+
+		// No cached response, return the original response and error
+		return response, err
+	}
+
+	replayer.padlock.Lock()
+	defer replayer.padlock.Unlock()
+
+	// We have a response, cache it and return it
+	replayer.gets[request.URL.String()] = response
+	return response, nil
+}
+
+func (replayer *replayRoundTripper) roundTripPut(request *http.Request) (*http.Response, error) {
+	// Calculate a hash of the request body to use as a cache key
+	// We need this whether we are updating our cache or replaying
+	hash := replayer.hashOfPUTBody(request)
+
+	response, err := replayer.inner.RoundTrip(request)
+	if err != nil {
+		// We have an error - return it, unless it's from go-vcr
+		if !errors.Is(err, cassette.ErrInteractionNotFound) {
+			return response, err
+		}
+
+		replayer.padlock.Lock()
+		defer replayer.padlock.Unlock()
+
+		// We didn't find an interaction, see if we have a cached response to return
+		if cachedResponse, ok := replayer.puts[hash]; ok {
+			// Remove it from the cache to ensure we only replay it once
+			delete(replayer.puts, hash)
+			replayer.log.Info("Replaying PUT request", "url", request.URL.String(), "hash", hash)
+			return cachedResponse, nil
+		}
+
+		// No cached response, return the original response and error
+		return response, err
+	}
+
+	replayer.padlock.Lock()
+	defer replayer.padlock.Unlock()
+
+	// We have a response, cache it and return it
+	replayer.puts[hash] = response
+	return response, nil
+}
+
+// hashOfPUTBody calculates a hash of the body of a PUT request, for use as a cache key.
+// The body is santised before calculating the hash to ensure that the same request body always results in the same hash.
+func (replayer *replayRoundTripper) hashOfPUTBody(request *http.Request) string {
```

Comment:
You could possibly just call this `hashOfBody`

Reply:
Good idea.